File: app/src/main/java/com/example/narthana/genpass/MultiSelectMultiListPreference.java
package com.example.narthana.genpass;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.os.Parcel;
import android.os.Parcelable;
import android.preference.DialogPreference;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;

import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * Created by narthana on 6/01/17.
 */

public class MultiSelectMultiListPreference extends DialogPreference
{
    private static final int DEFAULT_NUM_CHECKS = 1;

    private final CharSequence[] mEntries;
    private final CharSequence[] mEntryValues;
    private final CharSequence[] mColumnEntries;
    private final CharSequence[] mColEntryValues;
    private final CharSequence[] mColumnDependencies;

    private final int numRows;
    private final int numCols;

    private CheckBox[][] mCheckBoxes;
    private Map<String, Set<String>> mValues;
    private Map<String, Set<String>> mSelectedValues;

//    private boolean mPreferenceChanged;

    public MultiSelectMultiListPreference(Context context, AttributeSet attrs)
    {
        super(context, attrs);

        setPersistent(true);
        final TypedArray a = context.getTheme().obtainStyledAttributes(
                attrs,
                R.styleable.MultiSelectMultiListPreference,
                0,
                0
        );

        mEntries = a.getTextArray(R.styleable.MultiSelectMultiListPreference_android_entries);
        mEntryValues = a.getTextArray(R.styleable.MultiSelectMultiListPreference_android_entryValues);
        mColumnEntries = a.getTextArray(R.styleable.MultiSelectMultiListPreference_columnEntries);
        mColEntryValues = a.getTextArray(R.styleable.MultiSelectMultiListPreference_columnEntryValues);
        mColumnDependencies = a.getTextArray(R.styleable.MultiSelectMultiListPreference_columnDependencies);
        a.recycle();

        numRows = mEntries.length;
        numCols = mColumnEntries.length;
    }

    @Override
    protected View onCreateDialogView()
    {
        // create a table with a grid of checkboxes
        TableLayout table = new TableLayout(getContext());
        table.setLayoutParams(new ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT
        ));

        // the 0th column should stretch to fill the dialog
        table.setColumnStretchable(0, true);
        int padding = Utility.dpToPx(getContext(), 14);
        table.setPaddingRelative(padding, padding, padding, padding);

        // create header row
        TableRow header = new TableRow(getContext());

        // params for the first row
        TableLayout.LayoutParams fistRowParams = new TableLayout.LayoutParams(
                TableRow.LayoutParams.MATCH_PARENT,
                TableRow.LayoutParams.WRAP_CONTENT
        );
        table.addView(header, fistRowParams);

        int pad = Utility.dpToPx(getContext(), 3);
        // loop to create the column headings and put them in the header row
        for (int j = 0; j < numCols;)
        {
            TextView headingTextView = new TextView(getContext());
            headingTextView.setText(mColumnEntries[j]);
            headingTextView.setPadding(pad, pad, pad, pad);
            TableRow.LayoutParams headingParams = new TableRow.LayoutParams(
                    TableRow.LayoutParams.MATCH_PARENT,
                    TableRow.LayoutParams.WRAP_CONTENT
            );
            headingParams.gravity = Gravity.CENTER_HORIZONTAL;
            headingParams.column = ++j;
            header.addView(headingTextView, headingParams);
        }

        // create the rest of rows
        TableLayout.LayoutParams rowParams = new TableLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
        );

        // we can recycles the params for the row labels, so take them out of the loop
        TableRow.LayoutParams entryParams = new TableRow.LayoutParams(
                TableRow.LayoutParams.MATCH_PARENT,
                TableRow.LayoutParams.WRAP_CONTENT
        );
        entryParams.gravity = Gravity.CENTER_VERTICAL;
        entryParams.column = 0;

        mCheckBoxes = new CheckBox[numRows][numCols];

        for (int i = 0; i < numRows; ++i)
        {
            TableRow row = new TableRow(getContext());

            // create the row label and add it to the row
            TextView entryTextView = new TextView(getContext());
            entryTextView.setText(mEntries[i]);
            row.addView(entryTextView, entryParams);

            // create the checkboxes
            for (int j = 0; j < numCols;)
            {
                final int iPos = i;
                final int jPos = j;

                CheckBox checkBox = new CheckBox(getContext());

//                mCheckBoxes.get(String.valueOf(j)).put(mEntryValues[i].toString(), checkBox);
                mCheckBoxes[i][j] = checkBox;

                checkBox.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener()
                {
                    @Override
                    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked)
                    {
                        String rowEntry = mEntryValues[iPos].toString();
//                        String colEntry = mColEntryValues[jPos].toString();
                        String colEntry = String.valueOf(jPos);

//                        mPreferenceChanged =
//                                !(mValues.get(colEntry).contains(rowEntry) == isChecked);

//                        if (mPreferenceChanged)
//                        {
                            if (isChecked) mSelectedValues.get(colEntry).add(rowEntry);
                            else mSelectedValues.get(colEntry).remove(rowEntry);
//                        }
                    }
                });

                TableRow.LayoutParams checkBoxParams = new TableRow.LayoutParams(
                        TableRow.LayoutParams.MATCH_PARENT,
                        TableRow.LayoutParams.WRAP_CONTENT
                );
                checkBoxParams.gravity = Gravity.CENTER_HORIZONTAL;
                checkBoxParams.column = ++j;

                row.addView(checkBox, checkBoxParams);
            }

            table.addView(row, rowParams);
        }

        return table;
    }

    @Override
    protected void onBindDialogView(View view)
    {
        super.onBindDialogView(view);
        updateCheckStates();
    }

    @Override
    protected void onDialogClosed(boolean positiveResult)
    {
//        if (positiveResult && mPreferenceChanged) persistValues(mSelectedValues);
//        if (!positiveResult) mSelectedValues = cloneValues(mValues);
//        mPreferenceChanged = false;

        if (positiveResult) persistValues(mSelectedValues);
        else mSelectedValues = cloneValues(mValues);
    }

    @Override
    protected void onSetInitialValue(boolean restorePersistedValue, Object defaultValue)
    {
        persistValues(restorePersistedValue ? getValuesFromResources(mValues)
                                            : (Map<String , Set<String>>) defaultValue);
    }

    private Map<String, Set<String>> getValuesFromResources(Map<String, Set<String>> defaultValue)
    {
        Map<String, Set<String>> values = new HashMap<>(numCols);

        SharedPreferences prefs = getSharedPreferences();
        for (int j = 0; j < numCols; ++j)
        {
            // Note: do not remove the cloning of the list variable. It is necessary for
            // new data to be written to the preferences. Android expects the output of
            // Preference.getStringSet to not be modified. Thus when it receives it back
            // in onDialogClosed, it just keeps the old data we got here
            // source: stackoverflow.com/questions/12528836/shared-preferences-only-saved-first-time
            // source: developer.android.com/reference/android/content/SharedPreferences.html
            //     "Objects that are returned from the various get methods must be treated as
            //      immutable by the application."

            final Set<String> list = prefs.getStringSet(
                    getKey() + String.valueOf(j),
                    defaultValue != null ? defaultValue.get(String.valueOf(j)) : null
            );
            if (list != null) values.put(String.valueOf(j), new HashSet<String>(list));
        }
        return values;
    }

    @Override
    protected Object onGetDefaultValue(TypedArray a, int index)
    {
        Resources res = getContext().getResources();
        Map<String, Set<String>> values = new HashMap<>();
        TypedArray array = res.obtainTypedArray(a.getResourceId(index, -1));

        for (int j = 0, n = array.length(); j < n; ++j)
        {
            final Set<String> result = new HashSet<>();
            CharSequence[] value = array.getTextArray(j);

            for (CharSequence c : value) result.add(c.toString());

            values.put(String.valueOf(j), result);
        }

        array.recycle();
        return values;
    }

    private void persistValues(Map<String, Set<String>> values)
    {
        mValues = values;
        mSelectedValues = cloneValues(values);
        if (shouldPersist())
        {
//            Log.d(getClass().getSimpleName(), "Saving values");
//            Log.d(getClass().getSimpleName(), valuesToString(values));
            SharedPreferences.Editor editor = getEditor();
            editor.putBoolean(getKey(), true);
            for (Map.Entry<String, Set<String>> entry : values.entrySet())
                editor.putStringSet(getKey() + entry.getKey(), entry.getValue());
            editor.apply();
        }
    }

    @Override
    protected Parcelable onSaveInstanceState()
    {
//        Log.d(getClass().getSimpleName(), "Saving State");
//        Log.d(getClass().getSimpleName(), "selected values = " + valuesToString(mSelectedValues));
//        Log.d(getClass().getSimpleName(), "values = " + valuesToString(mValues));
        final Parcelable superState = super.onSaveInstanceState();
//        // Check whether this Preference is persistent (continually saved)
//        // No need to save instance state since it's persistent,use superclass state
//        if (isPersistent()) return superState;

        // Create instance of custom BaseSavedState
        final SavedState myState = new SavedState(superState);
        // Set the state's value with the class member that holds current
        // setting value
        myState.values = mSelectedValues;
        return myState;
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state)
    {
//        Log.d(getClass().getSimpleName(), "Restoring state");
        // Check whether we saved the state in onSaveInstanceState
        if (state == null || !state.getClass().equals(SavedState.class))
        {
            // Didn't save the state, so call superclass
            super.onRestoreInstanceState(state);
            return;
        }

        // Cast state to custom BaseSavedState and pass to superclass
        SavedState myState = (SavedState) state;
        super.onRestoreInstanceState(myState.getSuperState());

        // Set this Preference's widget to reflect the restored state
        mSelectedValues = myState.values;
//        Log.d(getClass().getSimpleName(), "selected values = " + valuesToString(mSelectedValues));
//        Log.d(getClass().getSimpleName(), "values = " + valuesToString(mValues));
        updateCheckStates();
    }

    private static class SavedState extends BaseSavedState
    {
        // Member that holds the setting's value
        // Change this data type to match the type saved by your Preference
        Map<String, Set<String>> values;

        public SavedState(Parcelable superState)
        { super(superState); }

        public SavedState(Parcel source)
        {
            super(source);
            // Get the current preference's value
            int size = source.readInt();
            values = new HashMap<>(size);
            String[] keys = new String[size];
            source.readStringArray(keys);
            for (String key : keys)
            {
                int n = source.readInt();
                String[] value = new String[n];
                source.readStringArray(value);
                values.put(key, new HashSet<String>(Arrays.asList(value)));
            }
        }

        @Override
        public void writeToParcel(Parcel dest, int flags)
        {
            super.writeToParcel(dest, flags);
            // Write the preference's value
            Set<String> keySet = values.keySet();
            String[] keysArray = keySet.toArray(new String[keySet.size()]);
            dest.writeInt(keysArray.length);
            dest.writeStringArray(keysArray);
            for (String key : keysArray)
            {
                Set<String> value = values.get(key);
                dest.writeInt(value.size());
                dest.writeStringArray(value.toArray(new String[value.size()]));
            }
        }

        // Standard creator object using an instance of this class
        public static final Parcelable.Creator<SavedState> CREATOR =
                new Parcelable.Creator<SavedState>()
        {
            public SavedState createFromParcel(Parcel in) { return new SavedState(in); }
            public SavedState[] newArray(int size) { return new SavedState[size];}
        };
    }

    private void updateCheckStates()
    {
        if (mCheckBoxes != null)
            for (int i = 0; i < numRows; ++i) for (int j = 0; j < numCols; ++j)
                mCheckBoxes[i][j].setChecked(
                        mSelectedValues.get(String.valueOf(j))
                                .contains(mEntryValues[i].toString())
                );
    }

    private <K, T> Map<K, Set<T>> cloneValues(Map<K, Set<T>> values)
    {
        Map<K, Set<T>> newValues = new HashMap<>(values.size());
        for (Map.Entry<K, Set<T>> entry : values.entrySet())
            newValues.put(entry.getKey(), new HashSet<T>(entry.getValue()));
        return newValues;
    }

//    private String valuesToString(Map<String, Set<String>> values)
//    {
//        String keys = Arrays.toString(values.keySet().toArray());
//        StringBuilder valuesArrays = new StringBuilder();
//        for (Map.Entry<String, Set<String>> e : values.entrySet())
//            valuesArrays.append(Arrays.toString(e.getValue().toArray()));
//        return keys + valuesArrays.toString();
//    }
}

File: app/src/main/java/com/example/android/genpass/preference/MultiSelectMultiListPreference.kt
package com.example.android.genpass.preference

import android.content.Context
import android.content.res.TypedArray
import android.os.Parcel
import android.os.Parcelable
import android.preference.DialogPreference
import android.preference.Preference
import android.util.AttributeSet
import android.util.SparseArray
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.example.android.genpass.R

/**
 * Created by narthana on 6/01/17.
 */

class MultiSelectMultiListPreference(context: Context, attrs: AttributeSet):
        DialogPreference(context, attrs) {
    private val entries: List<String>
    private val entryValues: List<String>
    private val columnEntries: List<String>
    private lateinit var colEntryValues: List<String> // init in the hidden no arg constructor
    private val columnDeps: SparseArray<Set<Int>>
    private val checkBoxes: Array<Array<CheckBox>>

    private val numRows: Int
    private val numCols: Int

    private lateinit var prefValues: Map<String, Set<String>>
    private lateinit var selectedValues: Map<String, MutableSet<String>>

    init {
        val (entries, entryValues, columnEntries, colDepsId) = context.theme.obtainStyledAttributes(
                attrs,
                R.styleable.MultiSelectMultiListPreference,
                0, 0
        ).use {
            StyleAttrVals(
                    getTextArray(R.styleable.MultiSelectMultiListPreference_android_entries)
                            .map(CharSequence::toString),
                    getTextArray(R.styleable.MultiSelectMultiListPreference_android_entryValues)
                            .map(CharSequence::toString),
                    getTextArray(R.styleable.MultiSelectMultiListPreference_columnEntries)
                            .map(CharSequence::toString),
                    getResourceId(R.styleable.MultiSelectMultiListPreference_columnDependencies, -1)
            )
        }
        this.entries = entries
        this.entryValues = entryValues
        this.columnEntries = columnEntries

        numRows = this.entries.size
        numCols = this.columnEntries.size

        // gather the data on dependencies between columns
        columnDeps = context.resources.obtainTypedArray(colDepsId).use {
            SparseArray<Set<Int>>(length()).apply {
                (0 until length())
                        .map { context.resources.getStringArray(this@use.getResourceId(it, -1)) }
                        .map { it.map { colEntryValues.indexOf(it) } }
                        .forEach { put(it[0], it.drop(1).toSet()) }
            }
        }

        // create the check box 2d array
        checkBoxes = Array<Array<CheckBox>>(numRows) {
            Array<CheckBox>(numCols) { CheckBox(context) }
        }
    }

    override fun onCreateDialogView(): View {
        // create a table with a grid of checkboxes
        val table = TableLayout(context).apply {
            layoutParams = ViewGroup.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.MATCH_PARENT
            )
            setColumnStretchable(0, true) // the 0th column should stretch to fill the dialog
            val padding = DIALOG_PADDING.dpToPx(context)
            setPaddingRelative(padding, padding, padding, padding)
        }

        // create header row
        val header = TableRow(context)

        // params for the first row
        table.addView(header, TableLayout.LayoutParams(
                TableRow.LayoutParams.MATCH_PARENT,
                TableRow.LayoutParams.WRAP_CONTENT
        ))

        // create the column headings and put them in the header row
        val headingParams = TableRow.LayoutParams(
                TableRow.LayoutParams.MATCH_PARENT,
                TableRow.LayoutParams.WRAP_CONTENT
        ).apply { gravity = Gravity.CENTER_HORIZONTAL }

        val headingPad = HEADING_PADDING.dpToPx(context)
        for (j in 0 until numCols) header.addView(
                TextView(context).apply {
                    text = columnEntries[j]
                    setPadding(headingPad, headingPad, headingPad, headingPad)
                },
                headingParams.apply { column = j }
        )

        // create the rest of rows
        val rowParams = TableLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
        )
        val entryParams = TableRow.LayoutParams(
                TableRow.LayoutParams.WRAP_CONTENT,
                TableRow.LayoutParams.WRAP_CONTENT
        ).apply {
            gravity = Gravity.CENTER_VERTICAL
            column = 0
        }

        for (i in checkBoxes.indices) {
            val row = TableRow(context)

            // create the row label and add it to the row
            row.addView(TextView(context).apply { text = entries[i] }, entryParams)

            // create the checkboxes
            for (j in checkBoxes[i].indices) {
                val checkBox = checkBoxes[i][j]

                // Set the check change listener
                checkBox.setOnCheckedChangeListener(
                        columnDeps[j]?.run { IndependentCheckChangeListener(i, j, this) }
                                ?: DependentCheckChangeListener(i, j)
                )

                checkBox.removeFromParent()

                row.addView(
                        checkBox,
                        TableRow.LayoutParams(
                                TableRow.LayoutParams.WRAP_CONTENT,
                                TableRow.LayoutParams.WRAP_CONTENT
                        ).apply {
                            gravity = Gravity.CENTER_HORIZONTAL
                            column = j
                        }
                )
            }
            table.addView(row, rowParams)
        }

        return table
    }

    override fun onBindDialogView(view: View) {
        super.onBindDialogView(view)
        updateCheckStates(selectedValues)
    }

    override fun onDialogClosed(positiveResult: Boolean) {
        if (positiveResult) persistValues(selectedValues)
        else selectedValues = makeSelectable(prefValues)
    }

    override fun onSetInitialValue(restorePersistedValue: Boolean, defaultValue: Any?) {
        val def = defaultValue as? Map<String, Set<String>> ?: mapOf()
        persistValues(if (restorePersistedValue) getValuesFromResources(def) else def)
    }

    override fun onGetDefaultValue(a: TypedArray, index: Int): Map<String, Set<String>>
            = context.resources.obtainTypedArray(a.getResourceId(index, -1)).use {
        colEntryValues = getTextArray(0).map(CharSequence::toString)
        (1 until length()).associateBy({ colEntryValues[it - 1] }) {
            getTextArray(it).map(CharSequence::toString).toSet()
        }
    }

    // Note: do not remove the cloning of v. It is necessary for
    // new data to be written to the preferences. Android expects the output of
    // Preference.getStringSetPref to not be modified. Thus when it receives it back
    // in onDialogClosed, it just keeps the old data we got here
    // source: stackoverflow.com/questions/12528836/shared-preferences-only-saved-first-time
    // source: developer.android.com/reference/android/content/SharedPreferences.html
    //     "Objects that are returned from the various get methods must be treated as
    //      immutable by the application."
    private fun getValuesFromResources(defaultValue: Map<String, Set<String>>):
            Map<String, Set<String>> = (0 until numCols).map { colEntryValues[it] }.associateBy(::id) {
        sharedPreferences.getStringSet(key + it, defaultValue[it]) ?: setOf()
    }

    private fun persistValues(values: Map<String, Set<String>>) {
        prefValues = values
        selectedValues = makeSelectable(values)
        if (shouldPersist()) {
            editor.apply {
                putBoolean(key, true)
                values.forEach { (k, v) -> putStringSet(key + k, v) }
            }.apply()
        }
    }

    private fun updateCheckStates(values: Map<String, Set<String>>) {
        for (i in checkBoxes.indices) for (j in checkBoxes[i].indices)
            checkBoxes[i][j].isChecked = values[colEntryValues[j]]?.contains(entryValues[i]) ?: false
        for (k in 0 until columnDeps.size()) {
            val independentColNo = columnDeps.keyAt(k)
            val dependents = columnDeps.get(independentColNo)
            for (d in dependents) for (i in checkBoxes.indices)
                checkBoxes[i][d].isEnabled = checkBoxes[i][independentColNo].isChecked
        }
    }

    private fun <K, T> makeSelectable(values: Map<K, Set<T>>): Map<K, MutableSet<T>>
            = values.entries.associateBy ({ it.key }) {
        if (it.value.isEmpty()) mutableSetOf() else it.value as MutableSet<T>
    }

    // Create instance of custom BaseSavedState
    // Set the state's value with the class member that holds current setting value
    override fun onSaveInstanceState(): Parcelable
            = SavedState(super.onSaveInstanceState()).apply { values = selectedValues }

    override fun onRestoreInstanceState(state: Parcelable?) {
        // Check whether we saved the state in onSaveInstanceState
        if (state == null || state.javaClass != SavedState::class.java) {
            // Didn't save the state, so call superclass
            super.onRestoreInstanceState(state)
            return
        }

        // Cast state to custom BaseSavedState
        val myState = state as SavedState

        // Set this Preference's widget to reflect the restored state
        selectedValues = myState.values
        updateCheckStates(selectedValues)
        super.onRestoreInstanceState(myState.superState)
    }

    private class SavedState: Preference.BaseSavedState {
        lateinit var values: Map<String, MutableSet<String>>

        constructor(superState: Parcelable): super(superState)

        constructor(source: Parcel): super(source) {
            val keys = mutableListOf<String>()
            source.readStringList(keys)
            values = keys.associateBy(::id) {
                val value = mutableListOf<String>()
                source.readStringList(value)
                value.toMutableSet()
            }
        }

        override fun writeToParcel(dest: Parcel, flags: Int) {
            dest.writeStringList(values.keys.toList())
            for (v in values.values) dest.writeStringList(v.toList())
            super.writeToParcel(dest, flags)
        }

        companion object @JvmField val CREATOR = creator(::SavedState)
    }

    private open inner class DependentCheckChangeListener(val row: Int, val col: Int):
            CompoundButton.OnCheckedChangeListener {
        override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean)
        { if (isChecked) check(row, col) else uncheck(row, col) }

        private fun check(row: Int, col: Int)
                = selectedValues[colEntryValues[col]]?.add(entryValues[row])
        private fun uncheck(row: Int, col: Int)
                = selectedValues[colEntryValues[col]]?.remove(entryValues[row])
    }

    private inner class IndependentCheckChangeListener(row: Int, col: Int, val deps: Set<Int>):
            DependentCheckChangeListener(row, col) {
        override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean) {
            super.onCheckedChanged(buttonView, isChecked)
            checkBoxes[row].run {
                for (d in deps) {
                    get(d).isEnabled = isChecked
                    get(d).isChecked = false
                }
            }
        }
    }

    private data class StyleAttrVals(val entries: List<String>, val entryValues: List<String>,
                                     val columnEntries: List<String>, val columnDepsId: Int)

    companion object {
        private const val DIALOG_PADDING = 14.0f
        private const val HEADING_PADDING = 3.0f
    }
}